src/documentMetadataManager.ts

```diff
@@ -0,0 +1,97 @@
+import { Disposable, TextDocument, Uri, workspace } from "vscode";
+import { Logger } from "./logging";
+import { getResourceManager } from "./storage/resourceManager";
+import { UriMetadataMap } from "./storage/types";
+
+const logger = new Logger("documentMetadataManager");
+
+/** Manager for VS Code {@link TextDocument}s that tracks metadata across document lifecycle events */
+export class DocumentMetadataManager {
+  private resourceManager = getResourceManager();
+
+  disposables: Disposable[] = [];
+
+  private constructor() {
+    this.disposables.push(
+      workspace.onDidOpenTextDocument(this.handleDocumentOpen, this),
+      workspace.onDidSaveTextDocument(this.handleDocumentSave, this),
+      workspace.onDidCloseTextDocument(this.handleDocumentClose, this),
+      // TODO: add other listeners (rename, delete, move, etc.) as needed
+    );
+  }
+
+  private static instance: DocumentMetadataManager | null = null;
+  static getInstance(): DocumentMetadataManager {
+    if (!DocumentMetadataManager.instance) {
+      DocumentMetadataManager.instance = new DocumentMetadataManager();
+    }
+    return DocumentMetadataManager.instance;
+  }
+
+  private async handleDocumentOpen(document: TextDocument) {
+    logger.debug("document opened", {
+      uri: document.uri.toString(),
+    });
+  }
+
+  private async handleDocumentClose(document: TextDocument) {
+    logger.debug("document closed", {
+      uri: document.uri.toString(),
+    });
+  }
+
+  /**
+   * Handler for when a document is saved. This primarily checks when a `file` document is saved
+   * and attempts to migrate any metadata from an `untitled` document to the saved document.
+   *
+   * When an `untitled` document is saved, the following happens:
+   * 1. A (new) `file` document is opened
+   * 2. The `file` document is saved with the same content as the `untitled` document
+   * 3. The `untitled` document is closed
+   *
+   * Between steps 2 and 3, both documents are briefly available and have the same content, so we
+   * can use the metadata from the `untitled` document to set the metadata for the `file` document.
+   */
+  private async handleDocumentSave(document: TextDocument) {
+    logger.debug("document saved", {
+      uri: document.uri.toString(),
+    });
+    if (document.uri.scheme !== "file") {
+      return;
+    }
+
+    const allMetadata: UriMetadataMap = await this.resourceManager.getAllUriMetadata();
```

Comment:
I think this setting of individual keys within the map will need protection from current documents firing (imagine 'save all')?

Push this down into resourcemanager, give it a method that does this under the auspices of a concurrent lock?

Reply:
Follow-up from offline discussion/pairing: we're keeping this one as-is without the mutex, but making the `setAllUriMetadata` [private](https://github.com/confluentinc/vscode/pull/1595/commits/766371e4c6d398582a21bab43af79d9a021e36ad) since it's only called by ResourceManager methods already using the mutex